currency-rate/cmd/main.go

```diff
@@ -1,113 +1,180 @@
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pkg/cron"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pkg/database"
+	dbCfg "github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pkg/database/config"
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pkg/settings"
+	"github.com/VictoriaMetrics/metrics"
 )
 
-const defaultCronSpec = "0 0 12 * * *"
+const (
+	ccFrom = "USD"
+	ccTo   = "UAH"
+)
 
-func StartCron(spec string, f func()) *cron.Cron {
-	c := cron.New()
-	_, err := c.AddFunc(spec, f)
-	if err != nil {
-		slog.Error("failed to add cron job", slog.Any("error", err))
-		return nil
-	}
-	slog.Info("cron job added", slog.Any("spec", spec))
-	c.Start()
-	return c
-}
+var appConfig appCfg.Config
 
-func InitDatabase() (*database.DB, error) {
+func NewDatabase() (*database.DB, error) {
 	db, err := database.New(dbCfg.NewFromEnv())
 	if err != nil {
 		return nil, err
 	}
-	if err := db.Migrate(&models.User{}, &models.Rate{}); err != nil {
+	if err := db.Migrate(&models.User{}); err != nil {
 		return nil, err
 	}
+	slog.Debug("database created")
 	return db, nil
 }
 
-func InitFetchers() fetchers.RateFetcher {
-	// Initialize rate fetcher chain of responsibilities
+func NewLogger() *slog.Logger {
+	loggerOptions := &slog.HandlerOptions{}
+	if appConfig.Debug {
+		loggerOptions.Level = slog.LevelDebug
+	}
+	logger := slog.New(
+		slog.NewTextHandler(os.Stdout, loggerOptions),
+	).With(slog.Any("service", "currency-rate"))
+	return logger
+}
+
+func NewFetchers() []service.RateFetcher {
 	nbuFetcher := fetchers.NewNBURateFetcher()
 	currencyBeaconFetcher := fetchers.NewCurrencyBeaconFetcher(
-		os.Getenv("CURRENCY_BEACON_API_KEY"),
+		appConfig.CurrencyBeaconAPIKey,
 	)
-	currencyBeaconFetcher.SetNext(nbuFetcher)
-	return currencyBeaconFetcher
+	fetchers := []service.RateFetcher{currencyBeaconFetcher, nbuFetcher}
+	slog.Debug("fetchers initialized", slog.Any("fetchers", fetchers))
+	return fetchers
+}
+
+func NewCron(transportConfig transportCfg.Config, fetcher service.RateFetcher) *cron.Manager {
+	transportConfig.QueueName = appConfig.RateQueueName
+
+	producer, err := rateProducer.NewProducer(transportConfig)
+	if err != nil {
+		slog.Error("failed to create rate producer", slog.Any("error", err))
+		return nil
+	}
+	slog.Debug("rate producer created")
+
+	job := cronRate.NewCronJob(fetcher, producer, ccFrom, ccTo)
+	spec := appConfig.RateRefreshCronSpec
+	if spec == "" {
+		spec = "@every 5m"
+	}
+	slog.Debug("cron job spec", slog.Any("spec", spec))
+
+	cronManager := cron.NewManager()
+	err = cronManager.AddJob(spec, job.Run)
+	if err != nil {
+		slog.Error("adding cron job", slog.Any("error", err))
+		return nil
+	}
+	return cronManager
+}
+
+// Get preferred outbound ip of this machine.
+func getOutboundIP() string {
+	conn, err := net.Dial("udp", "8.8.8.8:80")
```

Comment:
does this piece of code try to get an external IP?

Reply:
yes, it does

Reply:
maybe rename it to getExternalIP or getPublicIP?